js/index.d.ts

```diff
@@ -0,0 +1,122 @@
+/// <reference lib="es2020" />
+
+declare interface MimeType {
+    mime: string;
+}
+
+export interface Resource {
+    name: string;
+    aliases: string[];
+    kind: MimeType | "template";
+    content: string;
+    dependencies?: string[];
+    permission?: number;
+}
+
+export interface FilterListMetadata {
+    homepage?: string;
+    title?: string;
+    expires?: number;
+    redirect?: string;
+}
+
+export interface ParseOptions {
+    format: typeof FilterFormat;
+    rule_types: typeof RuleTypes;
+}
+
+declare interface BlockerResult {
+    matched: boolean,
+    important: boolean,
+    redirect?: string,
+    rewritten_url?: string,
+    exception?: string,
+    filter?: string,
+}
+
+export class Engine {
+    constructor(rules: FilterSet, debug: boolean);
+    addResource(resource: Resource): boolean;
+    check(url: string, source_url: string, request_type: string, debug?: false): boolean;
+    check(url: string, source_url: string, request_type: string, debug: true): BlockerResult;
```

Comment:
Nit: we can pass an object (aka named parameters) instead of passing 4 parameters separately, and then we will have a better user experience.

Reply:
sounds like that's a breaking change as well, so I'll leave it as-is for now